sound/soc/intel/common/soc-acpi-intel-cml-match.c

```diff
@@ -9,45 +9,59 @@
 #include <sound/soc-acpi.h>
 #include <sound/soc-acpi-intel-match.h>
 
-static struct snd_soc_acpi_codecs cml_codecs = {
+static struct snd_soc_acpi_codecs rt5682_codecs = {
 	.num_codecs = 1,
 	.codecs = {"10EC5682"}
 };
 
-static struct snd_soc_acpi_codecs cml_spk_codecs = {
-	.num_codecs = 1,
-	.codecs = {"MX98357A"}
+static struct snd_soc_acpi_codecs rt5682_spk_codecs = {
+	.num_codecs = 2,
+	.codecs = {"10EC5682", "MX98357A"}
+};
+
+static struct snd_soc_acpi_codecs da7219_spk_codecs = {
+	.num_codecs = 2,
+	.codecs = {"10EC5682", "MX98357A"}
+};
+
+static struct snd_soc_acpi_codecs rt1011_spk_codecs = {
+	.num_codecs = 3,
+	.codecs = {"10EC1011", "10EC5682", "MX98357A"}
```

Comment:
and is this correct? Does this mean the ACPI match will fail if all 3 codecs arent present?

Reply:
if one of the listed codecs not present, the match will fail, and goes to the next item in mach table.

Reply:
@aiChaoSONG yes I understand that. My quest was whether this is an actual board config for CML?

Reply:
@aiChaoSONG the cml_rt1011_rt5682 machine driver doesnt seem to handle the MX98357A. Are you sure this is the driver to use with this config?

Reply:
@ranj063 the sof_rt5682 driver can handle two conditions: only rt5682 and rt5682 + max98357a, and the cml_rt1011_rt5682 driver uses the handler under sof_rt5682.

Reply:
@aiChaoSONG yes, then what about the combination of all three?

Reply:
@aiChaoSONG to make my question clearer, what issue are you fixing with this PR, on which board and with what codecs?

Reply:
@ranj063 it is a cml board with cml rt1011 and rt5682 codec, but due to a know issue of coreboot,  acpi id MX98357A is present even there is no max98357a codec on the hardware.
in our code base, MX98357A is ahead of 10EC1011, so MX98357A will be firstly matched, and a wrong codec driver will be loaded on this board, this will cause speaker not work, because rt1011 codec driver is not loaded. My first solution is to bring 10EC1011 ahead, but this seems not so good, so I take your suggestion and use machine_quirk and quirk_data to solve this. Even with machine_quirk and quirk_data, the order of mach table matters, too. Because we have three acpi IDs on this board, if MX98357A is present first, we will match to it, and load wrong dirver again.

Reply:
@aiChaoSONG Thanks for the explanation. Now I get the problem better but I'm not sure what the correct solution is. @plbossart we're going to need some input from you.

Reply:
i'd like a text explanation with pseudo code that shows which machine driver is selected. I don't have the time to reverse-engineer this stuff.

Reply:
@plbossart  @ranj063 This should be easier for you to understand
CML board 1: with *rt5682* and *max98357a*  codecs, ACPI IDs: **10EC5682** **MX98357A**
CML board 2: with *rt5682* and *rt1011*  codecs , ACPI IDs: **10EC5682**, **10EC1011** and **MX98357A**  (The MX98357A here is a coreboot issue, and may not be fixed)
```
// our current code
struct snc_soc_acpi_mach snd_soc_acpi_intel_cml_machines[] = {
    ...
    {
        .id = "MX98357A",             // board 1 should match this
        .drv_name = "sof_rt5682",
        ...
    },
    {
        .id = "10EC1011",              // board 2 should match this
        .drv_name = "cml_rt1011_rt5682",
        ...
    },
}
```
Both boards will match ***MX98357A***, this is correct for board 1, but wrong for board 2.
**Solution 1**: Bring 10EC1011 ahead, both boards can get the right match.
**Solution 2**: Bring 10EC1011 ahead and add machine_quirk field to ensure full codec list match. Because we have quirk_data defined currently, but it may not be used without machine_quirk.

***Note: the order of 10EC1011 and MX98357A still matters in solution 2***

@plbossart We are not sure which solution to choose, please give your suggestions.

Reply:
Solution c)
use .id=10EC5682 for both, and add a machine quirk with MX98357a for the first and 10EC1011 for the second.

That way you use the list of amplifiers to disambiguate the solution, and the order does not matter, nor does the presence of MAX98357a for the second case.

Reply:
@plbossart As I understand from your text, we keep two items both with .id=10EC5682,  and codec MX98357A in quirk_data of the first board, and 10EC1011 in the other. Do I get you right?

Codec list of board 2 is a super set of board 1, that means where board 1 can match, Board 2 matches, too.
Solution d): Let me push coreboot team to fix their issue, and eliminate MX98357A in board 2.

Reply:
@aiChaoSONG @plbossart 

Solution from Pierre will work , just that we need to swap the order
use .id=10EC5682 for both, and add a machine quirk with 10EC1011 for the first and MX98357a for the second.

To Answer @plbossart  further,

CML device #1 : match RT1011 and RT5682
CML device #2: match max98357a and RT5682

If we use #2 as 1st entry, device #1 also pick that because

In Coreboot, Device #2 is used as base config and max98357a and RT5682 added.
Device #1 adds a override to add RT1011. Now since max98367a is not an i2c device, it continue to be present in ACPI.

Hence device #1 ACPI contains max98357a . ( This is what @aiChaoSONG reported and i'll check with Coreboot if possible)

If we have the order as below it can solve for all, let me know if you are ok

1st entry : RT5682 ( quirk of RT1011)
2nd entry: RT5682 (quirt of MAX98357a)
3rd entry:  RT5682 ( no quirk - for sof_rt5682)
4th entry:  da7219 (quirk of MAX98357a)

let me know if the match works for all with this.

Reply:
@plbossart @sathyap-chrome The order matters even with Pierre's solution, right? I am fine with this, but Pierre is the one who makes the final decision.

Reply:
yes, this is what I had in mind.

1st entry : RT5682 ( quirk of RT1011)
2nd entry: RT5682 (quirt of MAX98357a)
3rd entry: RT5682 ( no quirk - for sof_rt5682)
4th entry: da7219 (quirk of MAX98357a)

I am not sure why I wrote the order does not matter, it does indeed. the point was to use the speaker amplifier that's really present on the board to remove the ambiguity.

We'd of course need a comment to state that the order does matter and why, so that future generations don't break this by accident.

Reply:
@plbossart repushed, and tested on board with max98357a and boad with rt1011, correct driver is loaded on both boards, please review again.